core/modules/util/ThreadPool.cc

```diff
@@ -258,4 +289,35 @@ void ThreadPool::waitForResize(int millisecs) {
     }
 }
 
+void ThreadPool::_incrPoolThreadCount() {
+    std::lock_guard<std::mutex> lockPool(_mxPool);
+    ++_poolThreadCount;
+    LOGS(_log, LOG_LVL_WARN, "&&&WARN incr _poolThreadCount=" << _poolThreadCount);
+}
+
+void ThreadPool::_decrPoolThreadCount() {
+    {
+        std::lock_guard<std::mutex> lockPool(_mxPool);
+        --_poolThreadCount;
+    }
+    LOGS(_log, LOG_LVL_WARN, "&&&WARN decr _poolThreadCount=" << _poolThreadCount);
+    _cvPool.notify_all(); // There should not be many threads waiting.
+}
+
+void ThreadPool::_waitIfAtMaxThreadPoolCount() {
+    std::unique_lock<std::mutex> lockPool(_mxPool);
+    LOGS(_log, LOG_LVL_WARN, "&&&WARN wait A _poolThreadCount=" << _poolThreadCount);
+    _cvPool.wait(lockPool, [this](){return (_poolThreadCount <= _maxThreadCount);});
+    LOGS(_log, LOG_LVL_WARN, "&&&WARN wait B _poolThreadCount=" << _poolThreadCount);
+}
+
+bool ThreadPool::atMaxThreadPoolCount() {
+    std::unique_lock<std::mutex> lockPool(_mxPool);
+    bool atMax = _poolThreadCount > _maxThreadCount;
+    if (atMax) {
+        LOGS(_log, LOG_LVL_WARN, "atMaxThreadPoolCount current=" << _poolThreadCount << " max=" << _maxThreadCount);
```

Comment:
Same comments as before on reporting as `WARN` instead of `DEBUG` or `TRACE`.

Reply:
I think it is important to know that the system has run out of threads for the pool. Currently, this should be a rare occurrence as is really not easy to test under load. If it isn't a rare thing, we need to know that too.